statistics/handle/storage/read.go

```diff
@@ -562,12 +593,32 @@ func loadNeededIndexHistograms(sctx sessionctx.Context, statsCache *cache.StatsC
 	if !ok {
 		return nil
 	}
+	var (
+		hgMeta         *statistics.Histogram
+		flag           int64
+		lastAnalyzePos *types.Datum
+		idxInfo        *model.IndexInfo
+	)
+	statsVer := int64(-1)
 	index, ok := tbl.Indices[idx.ID]
-	if !ok {
+	if (!ok && !tbl.ColAndIndexExistenceMap.Has(idx.ID, true)) || (ok && index.IsFullLoad()) {
 		statistics.HistogramNeededItems.Delete(idx)
 		return nil
 	}
-	hg, err := HistogramFromStorage(sctx, idx.TableID, index.ID, types.NewFieldType(mysql.TypeBlob), index.Histogram.NDV, 1, index.LastUpdateVersion, index.NullCount, index.TotColSize, index.Correlation)
+	if ok {
+		hgMeta = &index.Histogram
+		idxInfo = index.Info
+		flag = index.Flag
+		lastAnalyzePos = new(types.Datum)
+		lastAnalyzePos.SetUint64(index.LastUpdateVersion)
+	} else {
+		hgMeta, lastAnalyzePos, statsVer, flag, err = HistMetaFromStorage(sctx, &idx, nil)
+		if err != nil {
+			return err
```

Comment:
should add `statistics.HistogramNeededItems.Delete` here like above？

Reply:
Other places that meet errors also do not delete it from the map. I think this is the original behavior.